File: plugins/rd/skills/10-stages-developing/templates/java/FunctionTemplateTest.java
package com.example.module;

import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.*;

import java.util.Map;

/**
 * Tests for FunctionTemplate
 */
public class FunctionTemplateTest {

    private FunctionTemplate functionTemplate;

    @Before
    public void setUp() {
        functionTemplate = new FunctionTemplate();
    }

    @Test
    public void testFunctionNameSmoke() {
        Map<String, Object> result = functionTemplate.functionName("test", 10);
        assertNotNull(result);
        assertTrue(result.containsKey("key1"));
        assertTrue(result.containsKey("key2"));
    }

    @Test
    public void testFunctionNameValidInput() {
        Map<String, Object> result = functionTemplate.functionName("hello", 42);

        assertEquals("processed_hello", result.get("key1"));
        assertEquals(84, result.get("key2"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testFunctionNameEmptyString() {
        functionTemplate.functionName("", 10);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testFunctionNameNullString() {
        functionTemplate.functionName(null, 10);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testFunctionNameNegativeNumber() {
        functionTemplate.functionName("test", -1);
    }

    @Test
    public void testFunctionNameZero() {
        Map<String, Object> result = functionTemplate.functionName("test", 0);
        assertEquals(0, result.get("key2"));
    }

    @Test
    public void testFunctionNameLargeNumber() {
        Map<String, Object> result = functionTemplate.functionName("test", 1000000);
        assertEquals(2000000, result.get("key2"));
    }

    // Parametrized-style tests using test methods
    @Test
    public void testFunctionNameParametrized1() {
        testParametrizedCase("hello", 10, "processed_hello", 20);
    }

    @Test
    public void testFunctionNameParametrized2() {
        testParametrizedCase("world", 5, "processed_world", 10);
    }

    @Test
    public void testFunctionNameParametrized3() {
        testParametrizedCase("test", 0, "processed_test", 0);
    }

    private void testParametrizedCase(String inputStr, int inputNum,
                                       String expectedKey1, int expectedKey2) {
        Map<String, Object> result = functionTemplate.functionName(inputStr, inputNum);
        assertEquals(expectedKey1, result.get("key1"));
        assertEquals(expectedKey2, result.get("key2"));
    }
}

File: plugins/rd/skills/10-stages-developing/templates/go/function-template_test.go
package module

import (
	"strings"
	"testing"
)

func TestFunctionNameSmoke(t *testing.T) {
	result, err := FunctionName("test", 10)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Key1 == "" || result.Key2 == 0 {
		t.Error("expected non-zero result")
	}
}

func TestFunctionNameValidInput(t *testing.T) {
	result, err := FunctionName("hello", 42)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if result.Key1 != "processed_hello" {
		t.Errorf("expected Key1 to be 'processed_hello', got '%s'", result.Key1)
	}

	if result.Key2 != 84 {
		t.Errorf("expected Key2 to be 84, got %d", result.Key2)
	}
}

func TestFunctionNameEmptyString(t *testing.T) {
	_, err := FunctionName("", 10)

	if err == nil {
		t.Error("expected error for empty string")
	}

	if !strings.Contains(err.Error(), "empty") {
		t.Errorf("expected error to mention 'empty', got: %s", err.Error())
	}
}

func TestFunctionNameNegativeNumber(t *testing.T) {
	_, err := FunctionName("test", -1)

	if err == nil {
		t.Error("expected error for negative number")
	}

	if !strings.Contains(err.Error(), "non-negative") {
		t.Errorf("expected error to mention 'non-negative', got: %s", err.Error())
	}
}

func TestFunctionNameZero(t *testing.T) {
	result, err := FunctionName("test", 0)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if result.Key2 != 0 {
		t.Errorf("expected Key2 to be 0, got %d", result.Key2)
	}
}

func TestFunctionNameLargeNumber(t *testing.T) {
	result, err := FunctionName("test", 1000000)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if result.Key2 != 2000000 {
		t.Errorf("expected Key2 to be 2000000, got %d", result.Key2)
	}
}

// Table-driven tests
func TestFunctionNameParametrized(t *testing.T) {
	tests := []struct {
		name         string
		inputStr     string
		inputNum     int
		expectedKey1 string
		expectedKey2 int
		expectError  bool
	}{
		{"valid case 1", "hello", 10, "processed_hello", 20, false},
		{"valid case 2", "world", 5, "processed_world", 10, false},
		{"zero value", "test", 0, "processed_test", 0, false},
		{"empty string", "", 10, "", 0, true},
		{"negative number", "test", -1, "", 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := FunctionName(tt.inputStr, tt.inputNum)

			if tt.expectError {
				if err == nil {
					t.Error("expected error but got none")
				}
			} else {
				if err != nil {
					t.Fatalf("unexpected error: %v", err)
				}

				if result.Key1 != tt.expectedKey1 {
					t.Errorf("expected Key1='%s', got '%s'", tt.expectedKey1, result.Key1)
				}

				if result.Key2 != tt.expectedKey2 {
					t.Errorf("expected Key2=%d, got %d", tt.expectedKey2, result.Key2)
				}
			}
		})
	}
}
